File: modules/benchmarks/arrow-benchmarks-effects/src/jmh/kotlin/arrow/benchmarks/Async.kt
package arrow.benchmarks

import arrow.fx.IO
import arrow.fx.IODispatchers
import org.openjdk.jmh.annotations.Benchmark
import org.openjdk.jmh.annotations.CompilerControl
import org.openjdk.jmh.annotations.Fork
import org.openjdk.jmh.annotations.Measurement
import org.openjdk.jmh.annotations.Param
import org.openjdk.jmh.annotations.Scope
import org.openjdk.jmh.annotations.State
import org.openjdk.jmh.annotations.Warmup
import java.util.concurrent.TimeUnit

@State(Scope.Thread)
@Fork(2)
@Warmup(iterations = 10, time = 1, timeUnit = TimeUnit.SECONDS)
@Measurement(iterations = 10)
@CompilerControl(CompilerControl.Mode.DONT_INLINE)
open class Async {

  @Param("3000")
  var size: Int = 0

  private fun ioAsyncLoop(i: Int): IO<Int> =
    IO.unit.continueOn(IODispatchers.CommonPool).followedBy(
      if (i > size) IO.just(i) else ioAsyncLoop(i + 1)
    )

  @Benchmark
  fun io(): Int =
    ioAsyncLoop(0).unsafeRunSync()

  @Benchmark
  fun catsIO(): Int =
    arrow.benchmarks.effects.scala.cats.`Async$`.`MODULE$`.unsafeIOAsyncLoop(size, 0)

  @Benchmark
  fun scalazZIO(): Int =
    arrow.benchmarks.effects.scala.zio.`Async$`.`MODULE$`.unsafeIOAsyncLoop(size, 0)
}

File: modules/benchmarks/arrow-scala-benchmarks/src/main/scala/cats/Async.scala
package arrow.benchmarks.effects.scala.cats

import cats.effect.{ContextShift, IO}
import cats.implicits._

import scala.concurrent.ExecutionContext

object Async {

  implicit val cs: ContextShift[IO] = IO.contextShift(ExecutionContext.global)

  def ioAsyncLoop(size: Int, i: Int): IO[Int] =
    IO.shift *> (if (i > size) IO.pure(i) else ioAsyncLoop(size, i + 1))

  def unsafeIOAsyncLoop(size: Int, i: Int): Int =
    ioAsyncLoop(size, i).unsafeRunSync()

}
